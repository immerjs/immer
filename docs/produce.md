---
id: produce
title: Using produce
---

<div id="codefund"><!-- fallback content --></div>

<details>
    <summary style="color: white; background:#c200c2;padding:5px;margin:5px;border-radius:2px">egghead.io lesson 3: Simplifying deep updates with _produce_</summary>
    <br>
    <div style="padding:5px;">
        <iframe style="border: none;" width=760 height=427  src="https://egghead.io/lessons/javascript-simplify-deep-state-updates-using-immer-produce/embed" ></iframe>
    </div>
    <a style="font-style:italic;padding:5px;margin:5px;"  href="https://egghead.io/lessons/javascript-simplify-deep-state-updates-using-immer-produce">Hosted on egghead.io</a>
</details>

The Immer package exposes a default function that does all the work.

`produce(currentState, producer: (draftState) => void): nextState`

There is also a curried overload that is explained [in a later section](curried-produce).

## Example

```javascript
import produce from "immer"

const baseState = [
	{
		todo: "Learn typescript",
		done: true
	},
	{
		todo: "Try immer",
		done: false
	}
]

const nextState = produce(baseState, draftState => {
	draftState.push({todo: "Tweet about it"})
	draftState[1].done = true
})
```

The interesting thing about Immer is that the `baseState` will be untouched, but the `nextState` will reflect all changes made to `draftState`.

```javascript
// the new item is only added to the next state,
// base state is unmodified
expect(baseState.length).toBe(2)
expect(nextState.length).toBe(3)

// same for the changed 'done' prop
expect(baseState[1].done).toBe(false)
expect(nextState[1].done).toBe(true)

// unchanged data is structurally shared
expect(nextState[0]).toBe(baseState[0])
// changed data not (d√ªh)
expect(nextState[1]).not.toBe(baseState[1])
```
