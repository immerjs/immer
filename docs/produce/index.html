<!DOCTYPE html>
<html>
  <head>
    <meta charset="UTF-8">
    <meta http-equiv="refresh" content="0; url=/immer/produce">
    <link rel="canonical" href="/immer/produce" />
  </head>
  <script>
    window.location.href = '/immer/produce' + window.location.search + window.location.hash;
  </script>
</html>