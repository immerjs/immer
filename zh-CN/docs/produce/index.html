<!DOCTYPE html>
<html>
  <head>
    <meta charset="UTF-8">
    <meta http-equiv="refresh" content="0; url=/immer/zh-CN/produce">
    <link rel="canonical" href="/immer/zh-CN/produce" />
  </head>
  <script>
    window.location.href = '/immer/zh-CN/produce' + window.location.search + window.location.hash;
  </script>
</html>