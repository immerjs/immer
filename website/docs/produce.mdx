---
id: produce
title: Using produce
---

<center>
	<div
		data-ea-publisher="immerjs"
		data-ea-type="image"
		className="horizontal bordered"
	></div>
</center> <details>
	<summary className="egghead-summary">
		egghead.io lesson 3: Simplifying deep updates with _produce_
	</summary>
	<br />
	<div>
		<iframe
			width="760"
			height="427"
			scrolling="no"
			src="https://egghead.io/lessons/javascript-simplify-deep-state-updates-using-immer-produce/embed"
		></iframe>
	</div>
	<a
		className="egghead-link"
		href="https://egghead.io/lessons/javascript-simplify-deep-state-updates-using-immer-produce"
	>
		Hosted on egghead.io
	</a>
</details>

The Immer package exposes a default function that does all the work.

`produce(currentState, recipe: (draftState) => void): nextState`

`produce` takes a base state, and a _recipe_ that can be used to perform all the desired mutations on the `draft` that is passed in. The interesting thing about Immer is that the `baseState` will be untouched, but the `nextState` will reflect all changes made to `draftState`.

Inside the recipe, all standard JavaScript APIs can be used on the `draft` object, including field assignments, `delete` operations, and mutating array, Map and Set operations like `push`, `pop`, `splice`, `set`, `sort`, `remove`, etc.

Any of those mutations don't have to happen at the root, but it is allowed to modify anything anywhere deep inside the draft: `draft.todos[0].tags["urgent"].author.age = 56`

Note that the recipe function itself normally doesn't return anything. However, it is possible to return in case you want to replace the `draft` object in its entirety with another object, for more details see [returning new data](return).

## Example

```javascript
import produce from "immer"

const baseState = [
	{
		title: "Learn TypeScript",
		done: true
	},
	{
		title: "Try Immer",
		done: false
	}
]

const nextState = produce(baseState, draftState => {
	draftState.push({title: "Tweet about it"})
	draftState[1].done = true
})
```

```javascript
// the new item is only added to the next state,
// base state is unmodified
expect(baseState.length).toBe(2)
expect(nextState.length).toBe(3)

// same for the changed 'done' prop
expect(baseState[1].done).toBe(false)
expect(nextState[1].done).toBe(true)

// unchanged data is structurally shared
expect(nextState[0]).toBe(baseState[0])
// ...but changed data isn't.
expect(nextState[1]).not.toBe(baseState[1])
```

### Terminology

- `(base)state`, the immutable state passed to `produce`
- `recipe`: the second argument of `produce`, that captures how the base state should be "mutated".
- `draft`: the first argument of any `recipe`, which is a proxy to the original base state that can be safely mutated.
- `producer`. A function that uses `produce` and is generally of the form `(baseState, ...arguments) => resultState`

Note that it isn't strictly necessary to name the first argument of the recipe `draft`. You can name it anything you want, for example `users`. Using `draft` as a name is just a convention to signal: "mutation is OK here".
