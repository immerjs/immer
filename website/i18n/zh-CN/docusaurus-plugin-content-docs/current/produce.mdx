---
id: produce
title: 使用 produce
---

<center>
	<div
		data-ea-publisher="immerjs"
		data-ea-type="image"
		className="horizontal bordered"
	></div>
</center>
<details>
	<summary className="egghead-summary">
		egghead.io 第 3 课：使用 produce 简化深度更新
	</summary>
	<br />
	<div>
		<iframe
			width="760"
			height="427"
			scrolling="no"
			src="https://egghead.io/lessons/javascript-simplify-deep-state-updates-using-immer-produce/embed"
		></iframe>
	</div>
	<a
		className="egghead-link"
		href="https://egghead.io/lessons/javascript-simplify-deep-state-updates-using-immer-produce"
	>
		Hosted on egghead.io
	</a>
</details>

Immer 包暴露了一个完成所有工作的默认函数。

`produce(currentState, recipe: (draftState) => void): nextState`

`produce` 需要一个 `baseState`，以及一个可用于对传入的 `draft` 进行所有所需更改的 `recipe`。关于 Immer 的有趣之处在于 baseState 将保持不变，但 nextState 将反映对 DraftState 所做的所有更改.

在 `recipe` 中，所有标准的 JavaScript API 都可以在 `draft` 对象上使用，包括属性字段分配、删除操作和修改数组、Map 和 Set 操作，如 push、pop、splice、set、sort、remove 等。

这些 `mutations` 中的任何一个都不必发生在初始对象上，但它可以修改 `draft` 深处的任何内容：`draft.todos[0].tags["urgent"].author.age = 56`

请注意，`recipe` 函数通常不会返回任何内容。但是，如果您想用另一个对象完全替换 `draft`，则可以返回，有关更多详细信息，请参阅返回[新数据](return)。

## 例子

```javascript
import produce from "immer"

const baseState = [
	{
		title: "Learn TypeScript",
		done: true
	},
	{
		title: "Try Immer",
		done: false
	}
]

const nextState = produce(baseState, draftState => {
	draftState.push({title: "Tweet about it"})
	draftState[1].done = true
})
```

```javascript
// 新的 item 仅仅被添加到了 next state
// base state 没有被修改
expect(baseState.length).toBe(2)
expect(nextState.length).toBe(3)

// 同上
expect(baseState[1].done).toBe(false)
expect(nextState[1].done).toBe(true)

// 未修改的数据结构共享
expect(nextState[0]).toBe(baseState[0])
// 改变的数据不是
expect(nextState[1]).not.toBe(baseState[1])
```

### 术语

- `(base)state`， 传递给 `produce` 的不可变状态
- `recipe`: `produce` 的第二个参数，它捕获了 base state 应该如何 `mutated`。
- `draft`: 任何 `recipe` 的第一个参数，它是可以安全 `mutate` 的原始状态的代理。
- `producer`. 一个使用 `produce` 的函数，通常形式为 `(baseState, ...arguments) => resultState`

请注意，命名 `recipe` 的第一个参数 `draft` 并不是绝对必要的。您可以将其命名为任何您想要的名称，例如 `user`。使用 `draft` 作为名称只是一个约定，以表明：“这里的 `mutation` 是可以的”。
