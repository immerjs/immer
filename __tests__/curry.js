"use strict"
import produce, {setUseProxies} from "../src/index"

runTests("proxy", true)
runTests("es5", false)

function runTests(name, useProxies) {
    describe("curry - " + name, () => {
        setUseProxies(useProxies)

        it("should check arguments", () => {
            expect(() => produce()).toThrowErrorMatchingSnapshot()
            expect(() => produce({})).toThrowErrorMatchingSnapshot()
            expect(() => produce({}, {})).toThrowErrorMatchingSnapshot()
            expect(() =>
                produce({}, () => {}, [])
            ).toThrowErrorMatchingSnapshot()
        })

        it("should support currying", () => {
            const state = [{}, {}, {}]
            const mapper = produce((item, index) => {
                item.index = index
            })

            expect(state.map(mapper)).not.toBe(state)
            expect(state.map(mapper)).toEqual([
                {index: 0},
                {index: 1},
                {index: 2}
            ])
            expect(state).toEqual([{}, {}, {}])
        })

        it("should support returning new states from curring", () => {
            const reducer = produce((item, index) => {
                if (!item) {
                    return {hello: "world"}
                }
                item.index = index
            })

            expect(reducer(undefined, 3)).toEqual({hello: "world"})
            expect(reducer({}, 3)).toEqual({index: 3})
        })

        it("should support passing an initial state as second argument", () => {
            const reducer = produce(
                (item, index) => {
                    item.index = index
                },
                {hello: "world"}
            )

            expect(reducer(undefined, 3)).toEqual({hello: "world", index: 3})
            expect(reducer({}, 3)).toEqual({index: 3})
            expect(reducer()).toEqual({hello: "world", index: undefined})
        })

        it("can has fun with change detection", () => {
            const spread = produce(Object.assign)

            const base = {
                x: 1,
                y: 1
            }

            expect({...base}).not.toBe(base)
            expect(spread(base, {})).toBe(base)
            expect(spread(base, {y: 1})).toBe(base)
            expect(spread(base, {...base})).toBe(base)
            expect(spread(base, {...base, y: 2})).not.toBe(base)
            expect(spread(base, {...base, y: 2})).toEqual({x: 1, y: 2})
            expect(spread(base, {z: 3})).toEqual({x: 1, y: 1, z: 3})
            expect(spread(base, {y: 1})).toBe(base)
        })
    })
}
