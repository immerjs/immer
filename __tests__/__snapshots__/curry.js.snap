// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`curry - proxy > should check arguments 1`] = `[Error: [Immer] The first or second argument to \`produce\` must be a function]`;

exports[`curry - proxy > should check arguments 2`] = `[Error: [Immer] The first or second argument to \`produce\` must be a function]`;

exports[`curry - proxy > should check arguments 3`] = `[Error: [Immer] The first or second argument to \`produce\` must be a function]`;

exports[`curry - proxy > should check arguments 4`] = `[Error: [Immer] The third argument to \`produce\` must be a function or undefined]`;
